Authentik Github SSO

uptownhr opened the pull request:
<!-- ELLIPSIS_HIDDEN -->


> [!IMPORTANT]
> Introduces `authentik_github_sso` module for Authentik-GitHub SAML SSO integration, focusing on authentication without user provisioning.
> 
>   - **New Module**:
>     - Adds `authentik_github_sso` module for integrating Authentik with GitHub SAML SSO.
>     - Handles authentication only; does not support user provisioning or de-provisioning.
>     - Requires GitHub Enterprise plan.
>   - **Configuration**:
>     - Includes `main.tf`, `vars.tf`, `outputs.tf`, and `config.yaml` for Terraform setup.
>     - Uses providers: `authentik`, `kubernetes`, `kubectl`, `random`, `tls`.
>   - **Documentation**:
>     - Adds `README.md` and `index.mdx` with setup guide and usage instructions.
>     - Updates `changelog/edge.mdx` and `modules.json` to include the new module.
>   - **Setup Guide**:
>     - Provides step-by-step instructions for configuring SAML SSO at the enterprise level.
>     - Details on deploying GitHub provider and application in Authentik.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=Panfactum%2Fstack&utm_source=github&utm_medium=referral)<sup> for 9ff13b2cd9be6782239f9e66111a1b0f204a67a2. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
````diff
@@ -0,0 +1,68 @@
+   ```
+   ![SAML Form](doc_images/github-saml-form.png)
+7. Keep `Require SAML SSO authentication for all members ...` unchecked.
+8. Click on Save.
````
Save should be "Save" or `Save`

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
```diff
@@ -0,0 +1,68 @@
+   ![SAML Form](doc_images/github-saml-form.png)
+7. Keep `Require SAML SSO authentication for all members ...` unchecked.
+8. Click on Save.
+9. Note the `assertion consumer service URL` and `organization single sign-on URL`. We will need it next.
```
Is the capitalization here correct?

uptownhr replied:
Yes

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
```diff
@@ -0,0 +1,68 @@
+   ![SAML Form](doc_images/github-saml-form.png)
+7. Keep `Require SAML SSO authentication for all members ...` unchecked.
+8. Click on Save.
+9. Note the `assertion consumer service URL` and `organization single sign-on URL`. We will need it next.
```
"We will need these later."

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
```diff
@@ -0,0 +1,68 @@
+![Audience and ACS URL](doc_images/github-audience-acs-url.png)
+
+1. Copy and set the `assertion consumer service URL` from Github to the `acs_url` input.
+2. Copy and set the `organization single sign-on URL` but without `/sso` from Github to the `audience` input.
```
Provide an example. "For example, if `organization single sign-on URL` is `abc.com/sso`, you should paste `abc.com`."

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
```diff
@@ -0,0 +1,68 @@
+
+Once SSO integration is confirmed to be working, it is recommended to enable `Require SAML SSO authentication`. 
+![Require SAML SSO Authentication](doc_images/github-require-saml-sso.png) 
+Not enabling this setting will prevent centralized access control through Authentik.  
```
Replace: "Leaving this disabled with negate any benefit of the SSO integration with Authentik."

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
````diff
@@ -0,0 +1,68 @@
+   -----END CERTIFICATE-----
+   ```
+   ![SAML Form](doc_images/github-saml-form.png)
+7. Keep `Require SAML SSO authentication for all members ...` unchecked.
````
I see it is from the screenshot. Resolved.

fullykubed commented on packages/infrastructure/authentik_github_sso/vars.tf:
```diff
@@ -0,0 +1,43 @@
+}
+
+variable "audience" {
+  description = "The Entity URL provided by the Service Provider"
```
You've said in the docs that this should not have `/sso` in it. I think it'd be appropriate to add a validation here as this seems like something someone could get wrong and then have it not work.

fullykubed commented on packages/infrastructure/authentik_github_sso/vars.tf:
```diff
@@ -0,0 +1,43 @@
+variable "ui_description" {
+  description = "The description to display in the Authentik web dashboard"
+  type        = string
+  default     = "Github"
```
`GitHub`

fullykubed commented on packages/infrastructure/authentik_github_sso/vars.tf:
```diff
@@ -0,0 +1,43 @@
+}
+
+variable "extra_allowed_groups" {
+  description = "Additional groups that can access Github"
```
`GitHub`

fullykubed commented on packages/infrastructure/authentik_github_sso/vars.tf:
```diff
@@ -0,0 +1,43 @@
+}
+
+variable "authentik_domain" {
+  description = "The domain name of the authentik instance"
```
`Authentik`

fullykubed commented on packages/infrastructure/authentik_github_sso/vars.tf:
```diff
@@ -0,0 +1,43 @@
+}
+
+variable "acs_url" {
+  description = "The ACS url provided by Github when configuring an external identity provider"
```
`GitHub`

fullykubed commented on packages/infrastructure/authentik_github_sso/vars.tf:
```diff
@@ -0,0 +1,43 @@
+variable "media_configmap" {
+  description = "The configmap holding the static media that Authentik will use"
```
`ConfigMap`

fullykubed commented on packages/infrastructure/authentik_github_sso/outputs.tf:
```diff
@@ -0,0 +1,14 @@
+}
+
+output "issuer_url" {
+  description = "The Authentik issuer URL for the Github provider"
```
`GitHub`

fullykubed commented on packages/reference/environments/production/us-east-2/authentik_github_sso/terragrunt.hcl:
```diff
@@ -0,0 +1,26 @@
+  authentik_domain  = dependency.kube_authentik.outputs.domain
+}
+
+skip = true
```
Don't forget to remove this once we have enterprise.

fullykubed commented on packages/website/src/content/docs/changelog/edge.mdx:
```diff
@@ -21,7 +21,8 @@ import MarkdownAlert from "@/components/markdown/MarkdownAlert.astro";
 ### Added
 
-* Adds [authentik_atlas_mongodb_sso](/* Adds [authentik_mongodb_atlas_sso](/docs/main/reference/infrastructure-modules/direct/authentik/authentik_mongodb_atlas_sso) & [mongodb_atlas_identity_provider](/docs/main/reference/infrastructure-modules/direct/authentik/mongodb_atlas_identity_provider) module for setting up the Authentik Application and Provider for Atlas MongoDB) module for setting up th ...
+* Adds [authentik_atlas_mongodb_sso](/docs/main/reference/infrastructure-modules/direct/authentik/authentik_mongodb_atlas_sso) & [mongodb_atlas_identity_provider](/docs/main/reference/infrastructure-modules/direct/authentik/mongodb_atlas_identity_provider) module for setting up the Authentik Application and Provider for Atlas MongoDB.
```
Links in the edge changelog should go to `/docs/edge` not `/docs/main`

fullykubed commented on packages/website/src/content/docs/changelog/edge.mdx:
```diff
@@ -21,7 +21,8 @@ import MarkdownAlert from "@/components/markdown/MarkdownAlert.astro";
 
-* Adds [authentik_atlas_mongodb_sso](/* Adds [authentik_mongodb_atlas_sso](/docs/main/reference/infrastructure-modules/direct/authentik/authentik_mongodb_atlas_sso) & [mongodb_atlas_identity_provider](/docs/main/reference/infrastructure-modules/direct/authentik/mongodb_atlas_identity_provider) module for setting up the Authentik Application and Provider for Atlas MongoDB) module for setting up th ...
+* Adds [authentik_atlas_mongodb_sso](/docs/main/reference/infrastructure-modules/direct/authentik/authentik_mongodb_atlas_sso) & [mongodb_atlas_identity_provider](/docs/main/reference/infrastructure-modules/direct/authentik/mongodb_atlas_identity_provider) module for setting up the Authentik Application and Provider for Atlas MongoDB.
+* Adds [authentik_github_sso](/docs/main/reference/infrastructure-modules/direct/authentik/authentik_github_sso) module for setting up the Authentik Application and Provider for Github.
```
`GitHub`

fullykubed commented on packages/infrastructure/authentik_github_sso/main.tf:
```diff
@@ -0,0 +1,150 @@
+  name = "superusers"
+}
+
+resource "authentik_policy_binding" "superuser_access" {
```
Why is superuser access different?

fullykubed replied:
It seems like this would be set up twice based on `local.combined_allowed_groups`

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
```diff
@@ -0,0 +1,68 @@
+
+# Authentik Github SSO
+
+This module configures Authentik for integration with Github SAML single sign-on.
```
`GitHub`

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
```diff
@@ -0,0 +1,68 @@
+
+### Setup Github SAML single sign-on
+
+1. Log in to Github and navigate to your organizations Settings.
```
`GitHub`

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
```diff
@@ -0,0 +1,68 @@
+### Test and Turn on Require SAML SSO
+
+1. Go to your Authentik instance.
+2. Find the Github application.
```
I think you should clarify that this is NOT the admin dashboard but the user dashboard (a screenshot might help). I was a little confused at first.

uptownhr commented on packages/infrastructure/authentik_github_sso/config.yaml:
```diff
@@ -0,0 +1,3 @@
+type: direct
+status: beta
```
@fullykubed let's connect on the terms here as that does not align with my understanding of `stable, beta, alpha`.

fullykubed replied:
See https://github.com/Panfactum/stack/blob/main/packages/website/src/components/markdown/StatusChip.astro

uptownhr replied:
resolved

uptownhr commented on packages/infrastructure/authentik_github_sso/main.tf:
```diff
@@ -0,0 +1,150 @@
+}
+
+locals {
+  combined_allowed_groups = toset([
```
what is expected? I did run lint just now, and nothing was called out or updated.

fullykubed replied:
IDK just looks weird to me so confirming you ran lint.

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
```diff
@@ -0,0 +1,68 @@
+3. Toggle on `Enable SAML authentication`.
+4. Set `Sign on URL` with the `url_sso_post` output value from above.
+5. Set `Issuer` with the `issuer_url` output value from above.
+6. Set `Public certificate` from the `saml_metadata` output value from above.
```
I don't understand why this would require XML parsing?

fullykubed replied:
You should be able to use an existing resource to output just the certificate. You shouldn't need to be manipulating the saml metadata at all.

uptownhr commented:
@fullykubed FYI this has been updated to integrate with the enterprise from organization.

fullykubed reviewed (changes requested):
I think this needs more updates if you are changing to the enterprise level protection. As in the docs need to be updated too.

uptownhr commented:
> I think this needs more updates if you are changing to the enterprise level protection. As in the docs need to be updated too.

@fullykubed the doc was updated to account for the enterprise experience. 

- updated to reference the enterprise dash
- removed steps that are no longer available in enterprise
- removed alert mentions of `enable saml for all members` as it is no longer an option

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
```diff
@@ -0,0 +1,53 @@
+
+### Start the GitHub SAML SSO Setup
+
+1. Log in to GitHub and navigate to your enterprise's dashboard. For example, Panfactum's enterprise url is https://github.com/enterprises/Panfactum.
```
Good to have an example, but personally it was extremely difficult to find out how to do this. I can imagine this being a blocker for less motivated folks. Can we add clarity on the recommended path to find this.

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
```diff
@@ -0,0 +1,53 @@
+  When a user is removed from Authentik, they will lose access to the organization. However, be aware of the following caveats:
+  - Any active session tokens that the user has with the GitHub web UI and PATs the user may have generated will not be automatically revoked. Until these tokens expire, the user may still interact with the web UI / API unless they are manually removed from the GitHub organization.
+</MarkdownAlert>
+
```
Add a MarkdownAlert warning that this sets up SAML at the enterprise level, NOT the organization level. Provide reasoning for why and explain the difference.

fullykubed commented on packages/infrastructure/authentik_github_sso/README.md:
```diff
@@ -0,0 +1,53 @@
+
+1. Go to your Authentik instance.
+2. Find the GitHub application. Ensure you are in the user dashboard, not the admin dashboard.
+   ![Github Application](doc_images/github-application.png)
```
GitHub

fullykubed commented on packages/infrastructure/authentik_github_sso/main.tf:
```diff
@@ -0,0 +1,148 @@
+}
+
+resource "authentik_application" "github" {
+  name              = "Github"
```
GitHub

fullykubed commented on packages/website/src/content/docs/main/reference/infrastructure-modules/direct/authentik/authentik_github_sso/index.mdx:
```diff
@@ -0,0 +1,160 @@
+
+1. Go to your Authentik instance.
+2. Find the GitHub application. Ensure you are in the user dashboard, not the admin dashboard.
+   ![Github Application](doc\_images/github-application.png)
```
GitHub

fullykubed commented on packages/infrastructure/authentik_github_sso/doc_images/github-audience-acs-url.png:
I am not sure what the security implications of posting a partial certificate are. Does that contain sensitive info?

fullykubed commented on packages/infrastructure/authentik_github_sso/doc_images/github-saml-form.png:
Potential sensitive info?

fullykubed commented on packages/reference/environments/panfactum.hcl:
```diff
@@ -292,12 +289,6 @@ generate "vault_provider" {
   }) : ""
 }
 
-generate "mongodb_atlas_provider" {
```
??

fullykubed commented on packages/website/src/content/docs/main/reference/infrastructure-modules/direct/authentik/authentik_github_sso/index.mdx:
```diff
@@ -0,0 +1,160 @@
+
+### saml\_metadata
+
+Description: The SAML metadata for the GitHub provider. https://en.wikipedia.org/wiki/SAML_metadata
```
Not a blocker, but for links you can use markdown format in the variable description so it isn't ugly like this.

itzsaga commented:
Testing Slack integration for @uptownhr.
